discovery/sync_manager.go

```diff
@@ -205,297 +186,301 @@ func (m *SyncManager) syncerHandler() {
+				close(newSyncer.doneChan)
+				continue
+			}
 
-	select {
-	case m.staleActiveSyncers <- &staleActiveSyncer{
-		syncer:       s,
-		transitioned: transitioned,
-		done:         done,
-	}:
-	case <-m.quit:
-	}
+			s := m.createGossipSyncer(newSyncer.peer)
 
-	// Before returning to the caller, we'll wait for the roundRobinHandler
-	// to signal us that the SyncManager has correctly updated its internal
-	// state after handling the stale active syncer.
-	select {
-	case <-done:
-	case <-m.quit:
-	}
-}
+			m.syncersMu.Lock()
+			switch {
+			// If we've exceeded our total number of active syncers,
+			// we'll initialize this GossipSyncer as passive.
+			case len(m.activeSyncers)+len(pendingActiveSyncers) >=
+				m.cfg.NumActiveSyncers:
+				s.setSyncType(PassiveSync)
+				m.inactiveSyncers[s.cfg.peerPub] = s
 
-// roundRobinHandler is the SyncManager's event loop responsible for managing
-// the round-robin queue of our active syncers to ensure they don't overlap and
-// request the same set of channels, which significantly reduces bandwidth
-// usage.
-//
-// NOTE: This must be run as a goroutine.
-func (m *SyncManager) roundRobinHandler() {
-	defer m.wg.Done()
+			// Otherwise, it should be initialized as active. If the
+			// initial historical sync has completed, then we can do
+			// so immediately.
+			case initialHistoricalSyncCompleted:
+				s.setSyncType(ActiveSync)
+				m.activeSyncers[s.cfg.peerPub] = s
+
+			// If we reach here, then we should initialize the
+			// GossipSyncer as active, but the initial historical
+			// sync hasn't completed, so we'll queue it up to
+			// transition once done.
+			default:
+				m.inactiveSyncers[s.cfg.peerPub] = s
+				pendingActiveSyncers[s.cfg.peerPub] = s
+			}
+			m.syncersMu.Unlock()
 
-	defer m.cfg.ActiveSyncerTimeoutTicker.Stop()
+			s.Start()
 
-	var (
-		// current will hold the current active syncer we're waiting for
-		// to complete its state machine.
-		current *GossipSyncer
-
-		// transitionNext will be responsible for containing the signal
-		// of when the current active syncer has completed its state
-		// machine. This signal allows us to transition the next pending
-		// active syncer, if any.
-		transitionNext chan struct{}
-	)
+			// Once we create the GossipSyncer, we'll signal to the
+			// caller that they can proceed since the SyncManager's
+			// internal state has been updated.
+			close(newSyncer.doneChan)
 
-	// transitionNextSyncer is a helper closure that we'll use to transition
-	// the next syncer queued up. If there aren't any, this will act as a
-	// NOP.
-	transitionNextSyncer := func() {
-		m.Lock()
-		current = m.nextPendingActiveSyncer()
-		m.Unlock()
-		for current != nil {
-			// Ensure we properly handle a shutdown signal.
-			select {
-			case <-m.quit:
-				return
-			default:
+			// We'll force a historical sync with the first peer we
+			// connect to, to ensure we get as much of the graph as
+			// possible.
+			if !attemptInitialHistoricalSync {
```

Comment:
looks like we can get rid of `pendingActiveSyncers` now. We can always add new syncers to the `inactiveSyncers` map, and let the syncer handler choose syncers from here after the historical sync is done :)

Reply:
Good point! Fixed.